lib/rules/no-inline-anonymous-functions-to-schedule-once.js

```diff
@@ -0,0 +1,64 @@
+'use strict';
+
+const utils = require('../utils/utils');
+const ERROR_MESSAGE = 'Do not use anonymous functions as arguments to `scheduleOnce` or `once`.';
+
+module.exports = {
+  ERROR_MESSAGE,
+
+  meta: {
+    docs: {
+      description:
+        'Disallows inline anonymous functions as arguments to run.scheduleOnce and run.once',
+      category: 'Possible Errors',
+      recommended: false,
+    },
+  },
+
+  create(context) {
+    function checkArgumentsForInlineFunction(node) {
+      node.arguments.forEach((argument, index) => {
+        if (utils.isFunctionExpression(argument)) {
+          context.report(node.arguments[index], ERROR_MESSAGE);
+        }
+      });
+    }
+
+    const allDedupingRunMethodNames = ['scheduleOnce', 'once'];
+    let importedRun = false;
+    let importedDedupingMethodNames = [];
+
+    return {
+      ImportDeclaration(node) {
+        if (node.source.value === '@ember/runloop') {
+          const namedImports = node.specifiers.map(specifier => {
+            return specifier.imported.name;
+          });
+
+          importedDedupingMethodNames = namedImports.filter(key => {
+            return allDedupingRunMethodNames.includes(key);
+          });
+
+          if (namedImports.includes('run')) {
+            importedRun = true;
+          }
+        }
+      },
+      CallExpression(node) {
+        if (!importedDedupingMethodNames.length && !importedRun) {
+          return;
+        }
+
+        if (importedDedupingMethodNames.includes(node.callee.name)) {
```

Comment:
How about restructuring this to more exactly check the parameters for each of the different functions and use helper functions:

```
if (isOnceCallWithAnonymousFunction(node) || isScheduleOnceCallWithAnonymousFunction(node)) {
  context.report(node.arguments[node.arguments.length - 1], ERROR_MESSAGE);
}

function isOnceCallWithAnonymousFunction(node) {
  return (
    isIdentifierOrMemberExpressionCall(node, 'run', 'once') &&
    node.arguments.length === 2 &&
    isAnonymousFunctionExpression(node.arguments[1])
  );
}

function isScheduleOnceCallWithAnonymousFunction(node) {
  return (
    isIdentifierOrMemberExpressionCall(node, 'run', 'scheduleOnce') &&
    node.arguments.length === 3 &&
    isAnonymousFunctionExpression(node.arguments[2])
  );
}

function isIdentifierOrMemberExpressionCall(node, objectName, propertyName) {
  return (
    (utils.isIdentifier(node.callee) && node.callee.name === propertyName) ||
    (utils.isMemberExpression(node.callee) &&
      utils.isIdentifier(node.callee.object) &&
      node.callee.object.name === objectName &&
      utils.identifier(node.callee.property) &&
      node.callee.property.name === propertyName)
  );
}

function isAnonymousFunctionExpression(node) {
  return utils.isFunctionExpression(node) || utils.isArrowFunctionExpression(node);
}

```

Reply:
Combining this style with the logic that detects which of the imports are available in the file gets messy and loses the readability that this snippet has.